src/pudl/analysis/demand_mapping.py

```diff
@@ -1,294 +1,942 @@
     """
-    Scales matrix by a vector with or without normalization.
+    Aggregates selected columns of the disaggregated layer based on arguments
 
-    Scales the linear mapping matrix by a particular variable e.g. If you want
-    to allocate demand by population, scale the area intersection fraction
-    matrix by the population of each of the FIPS using matrix_linear_scaling
-    once by axis_scale=1, appropriate dataframe, scale_col="POPULATION", then
-    allocate demand using matrix_linear_scaling once by axis_scale=0,
-    appropriate dataframe, scale_col="demand_mwh"
+    It is assumed that the data, which needs to be disaggregated, is present as
+    `constant` attributes in the GeoDataFrame. The data is mapped by the `by`
+    columns. So, first the data is disaggregated, according to the allocator
+    columns. Then, it is returned if aggregators list is empty. If it is not,
+    then the data is aggregated again to the aggregator level
 
     Args:
-        intersection_matrix (pandas.DataFrame): matrix with every one-one
-            mapping of two different geometry IDs
-        gdf_scale (pandas.DataFrame): dataframe with appropriate scaling data
-        gdf_scale_col (str): name of the column being allocated (needs same
-            name in dataframe and matrix)
-        axis_scale (int): axis of normalization and demand allocation
-            1 if data being multiplied to rows
-            0 if data being multiplied to columns
-        normalize (bool): normalize along the axis mentioned
+        disagg_layer (GeoDataframe): Completely disaggregated GeoDataFrame
+        by (str or list): single column or list of columns according to which
+        the constants to be allocated are mentioned (e.g. "Demand" (constant)
+        which needs to be allocated is mapped by "id". So, "id" is the `by`
+        column)
+        allocatees (str or list): single column or list of columns according to which
+        the constants to be allocated are mentioned (e.g. "Demand" (constant)
+        which needs to be allocated is mapped by "id". So, "demand" is the
+        `allocatees` column)
+        allocators (str or list): columns by which attribute is weighted and
+        allocated
+        aggregators (str or list): if empty list, the disaggregated data is
+        returned. If aggregators is mentioned, for example REEDs geometries, the
+        data is aggregated at that level.
 
     Returns:
-        pandas.DataFrame: Intersection matrix scaled by vector (row-wise or
-        column-wise)
-
+        geopandas.GeoDataFrame: Disaggregated GeoDataFrame with all the various
+        allocated demand columns, or aggregated by `aggregators`
     """
-    if axis_scale == 1:
-        unique_index = intersection_matrix.index
-        index_name = intersection_matrix.index.name
+    # Allowing for single and multiple allocators,
+    # aggregating columns and allocatees
+    if type(allocators) is not list:
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Using type() instead of isinstance() for a typecheck.](https://app.codacy.com/manual/zaneselvans/pudl/pullRequest?prid=5664147)